src/w3d/renderer/dazzle.h

```diff
@@ -14,21 +14,286 @@
+    virtual void Get_Obj_Space_Bounding_Box(AABoxClass &box) const override;
+    virtual void Scale(float scale) override { m_scale *= scale; }
+
+    void Set_Dazzle_Color(const Vector3 &col) { m_dazzleColor = col; }
+    void Set_Halo_Color(const Vector3 &col) { m_haloColor = col; }
+    void Set_Lensflare_Intensity(float intensity) { m_lensflareIntensity = intensity; }
+
+    unsigned int Get_Dazzle_Type() { return m_type; }
+
+    void Set_Layer(DazzleLayerClass *layer);
+
+#ifdef GAME_DLL
+    virtual const PersistFactoryClass &Get_Factory() const override;
+#endif
+
+    static void Set_Current_Dazzle_Layer(DazzleLayerClass *layer);
+
+    static void Init_Type(const DazzleInitClass &i);
+    static void Init_Lensflare(const LensflareInitClass &i);
     static void Init_From_INI(const INIClass *ini);
+    static unsigned int Get_Type_ID(const char *name);
+    static const char *Get_Type_Name(unsigned int id);
+    static DazzleTypeClass *Get_Type_Class(unsigned int id);
+    static unsigned int Get_Lensflare_ID(const char *name);
+    static LensflareTypeClass *Get_Lensflare_Class(unsigned int id);
+
     static void Deinit();
+    static void Install_Dazzle_Visibility_Handler(const DazzleVisibilityClass *visibility_handler);
+    static void Enable_Dazzle_Rendering(bool onoff) { s_dazzleRenderingEnabled = onoff; }
+    static bool Is_Dazzle_Rendering_Enabled() { return s_dazzleRenderingEnabled; }
+
+private:
+    void Vis_Render_Dazzle(SpecialRenderInfoClass &rinfo);
+    void Render_Dazzle(CameraClass *camera);
+
+    static bool s_dazzleRenderingEnabled;
+
+    DazzleRenderObjClass *m_succ;
+    unsigned int m_type;
+    float m_currentDazzleIntensity;
+    float m_currentDazzleSize;
+    float m_currentHaloIntensity;
+    float m_currentDistance;
+    Vector4 m_transformedLoc;
+    Vector3 m_currentVloc;
+    Vector3 m_currentDir;
+    Vector3 m_dazzleColor;
+    Vector3 m_haloColor;
+    float m_lensflareIntensity;
+    float m_scale;
+    float m_visibility;
+    bool m_onList;
+    float m_radius;
+    unsigned int m_creationTime;
+};
+
+class DazzleVisibilityClass
+{
+public:
+    virtual float Compute_Dazzle_Visibility(
+        RenderInfoClass &rinfo, DazzleRenderObjClass *dazzle, const Vector3 &point) const;
+};
+
+class DazzlePrototypeClass : public W3DMPO, public PrototypeClass
+{
+    IMPLEMENT_W3D_POOL(W3DMPO);
+
+public:
+    DazzlePrototypeClass() : m_dazzleType(0) {}
+    virtual ~DazzlePrototypeClass() override {}
+    virtual const char *Get_Name() const override { return m_name; }
+    virtual int Get_Class_ID() const override { return RenderObjClass::CLASSID_DAZZLE; }
+    virtual RenderObjClass *Create() override;
+    virtual void Delete_Self() override { delete this; };
+    W3DErrorType Load_W3D(ChunkLoadClass &cload);
+
+private:
+    StringClass m_name;
+    int m_dazzleType;
 };
 
 class DazzleLoaderClass : public PrototypeLoaderClass
 {
 public:
+    DazzleLoaderClass() {}
+    ~DazzleLoaderClass() {}
```

Comment:
Odd that prototype loader doesn't have a virtual destructor. This destructor will therefore likely never be called.

Reply:
Virtualness of functions from parent class should propagate to derived class, even when not specifying virtual keyword here.

Reply:
Yes and since the parent is not virtual this is not virtual. Just pointing out that it was odd. I've had a quick look and it's not UB though as loaders are static instances which are never deleted from their base pointer.

Reply:
Not calling a destructor is not Undefined Behavior, because it is very well defined what happens if not called. Whether or not that is bad depends on the context. If the Destructor has nothing to do, then there is logically no difference between being called or not called. The Compiler likely even makes it a no-op with Release.

Reply:
It is UB to delete a polymorphic type from its base pointer without a virtual destructor. Luckily unlike most items in Thyme we don't delete from a base pointer in this instance as all the classes are static instances and destructed from their derived type. Just because you think there is nothing to do doesn't change something being UB. 

```
std::vector<PrototypeLoader*> loaders;
for (auto* loader : loaders) delete loader;
```
This would be UB but we don't do that so it's fine.

Reply:
I guess that depends on your definition of Undefined Behavior. C++ is accurate about events in case of non virtual destructors, meaning the results are reproducible. Undefined Behavior for me means something is not defined and can have different outcomes, so for example an uninitialized integer could be 0 or 1354623. You would not know until you run the program.